generate-zephyr-dts.py

```diff
@@ -1,179 +1,136 @@
+# DTS handlers
+def disabled_handler(name, parm, csr):
+    return "    status = \"disabled\";\n"
 
 
-def ram_handler(region, **kw):
+def ram_handler(name, parm, csr):
+    return dts_reg(" ".join([
+                hex(csr['memories'][name]['base']),
+                hex(csr['memories'][name]['size'])]))
 
-    result = """
-&ram0 {{
-    reg = <0x{address} {size}>;
-}};
-""".format(address=hex(region['address'])[2:], size=hex(region['size']))
 
-    return result
+def ethmac_handler(name, parm, csr):
+    dtsi  = dts_reg(" ".join([
+                hex(csr['csr_bases'][name]),
+                hex(parm['size']),
+                hex(csr['memories'][name]['base']),
+                hex(csr['memories'][name]['size'])]))
+    dtsi += dts_intr(name, csr)
+    return dtsi
 
 
-def ethmac_handler(peripheral, **kw):
-    buf = kw['buffer']()
+def i2c_handler(name, parm, csr):
+    dtsi  = dts_reg(" ".join([
+                hex(csr['csr_bases'][name]),
+                hex(parm['size']),
+                hex(csr['csr_bases'][name] + parm['size']),
+                hex(parm['size'])]))
+    dtsi += dts_intr(name, csr)
+    return dtsi
 
-    result = """
-&{} {{
-""".format(kw['reference'])
 
-    result += """
-    reg = <{} {} {} {}>;
-""".format(hex(peripheral['address']),
-           hex(kw['size']),
-           hex(buf['address']),
-           hex(buf['size']))
+def peripheral_handler(name, parm, csr):
+    dtsi  = dts_reg(" ".join([
+                hex(csr['csr_bases'][name]),
+                hex(parm['size'])]))
+    try:
+        dtsi += dts_intr(name, csr)
+    except KeyError as e:
+        print('  dtsi key', e, 'not found, no interrupt override')
+    return dtsi
 
-    if 'constants' in peripheral:
-        if 'interrupt' in peripheral['constants']:
-            irq_no = int(peripheral['constants']['interrupt'], 0)
-            result += """
-    interrupts = <{} 0>;
-""".format(hex(irq_no))
 
-    result += """
-};
-"""
-    return result
-
-
-def i2c_handler(peripheral, **kw):
-    result = """
-&{} {{
-    reg = <{} {} {} {}>;
-}};
-""".format(kw['reference'],
-           hex(peripheral['address']),
-           hex(kw['size']),
-           hex(peripheral['address'] + kw['size']),
-           hex(kw['size']))
-
-    return result
-
-
-def peripheral_handler(peripheral, **kw):
-    result = """
-&{} {{
-""".format(kw['reference'])
-
-    result += """
-    reg = <{} {}>;
-""".format(hex(peripheral['address']),
-           hex(kw['size']))
-
-    if 'constants' in peripheral:
-        if 'interrupt' in peripheral['constants']:
-            irq_no = int(peripheral['constants']['interrupt'], 0)
-            result += """
-    interrupts = <{} 0>;
-""".format(hex(irq_no))
-
-    result += """
-};
-"""
-    return result
-
-
-peripheral_handlers = {
+overlay_handlers = {
     'uart': {
         'handler': peripheral_handler,
-        'reference': 'uart0',
-        'size': 0x18,
+        'alias': 'uart0',
+        'size': 0x20,
```

Comment:
Was the previous `size` value wrong? If so I suggest to extract this change into a separate commit.

Reply:
Just looking at the generated configuration from litex:

Last uart addres is:
`uart_rxfull  CSR_BASE + 0x1c`

So I expanded the IO region to include this register.